04_multilingual-ner.ipynb

```diff
@@ -15,16 +15,15 @@
   },
```

Comment:
<p>I think the order was ok before no? I know that it's a bit inconsistent that <code>plot_confusion_matrix</code> and <code>confusion_matrix</code> use reversed order of arguments but I think it's functionally correct, no?  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/nlp-with-transformers/notebooks/pull/76/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='72'/>

Reply:
<p>Thanks for double checking Leandro. I triple checked and I still believe the code in your notebook and the resulting confusion matrix are not correct.</p><p> </p><p>In your notebook you define plot_confusion_matrix as</p><pre class="ql-syntax" spellcheck="false">def plot_confusion_matrix(y_preds, y_true, labels):
</pre><p>where the order of the input is first predictions, then ground truth labels.</p><p> </p><p>However when you call the function you're passing the function inputs in the reverse order. First ground truth labels, then predicted labels:</p><pre class="ql-syntax" spellcheck="false">plot_confusion_matrix(
  df_tokens["labels"],
  df_tokens["predicted_label"],
  tags.names
)
</pre><p> </p><p>Hopefully with code what I'm trying to highlight is more clear.</p><p> </p><p>In any case, this is just one aspect that, in my opinion, makes the confusion matrix inaccurate. The second one is the mismatch between the values in the confusion matrix and the labels.</p><p> </p><p>If you run this simple code</p><pre class="ql-syntax" spellcheck="false">df_ILOC = df_tokens[df_tokens['labels'] == 'I-LOC']
(df_ILOC['labels'] == df_ILOC['predicted_label']).sum() / len(df_ILOC)
</pre><p>to check the accuracy of the <code>I-LOC</code> label, you will immediately notice that the model doesn't predict this label correctly 99% of the times as the confusion matrix produced by your notebook (and showed in the book) tells us. In fact, the model accuracy for this specific label (i.e. <code>I-LOC</code>) is around 85% and the actual label predicted correctly with 99% accuracy is <code>O</code>.</p><p> </p><p>The problems arises because sklearn confusion matrix sorts string inputs (i.e. our target and predicted labels) in alphabetical order. However, when you pass the labels to display here,</p><pre class="ql-syntax" spellcheck="false">disp = ConfusionMatrixDisplay(
  confusion_matrix=cm,
  display_labels=labels
)
</pre><p>you're passing then according to the order defined in the dataset which is not alphabetically</p><pre class="ql-syntax" spellcheck="false">['O', 'B-PER', 'I-PER', 'B-ORG', 'I-ORG', 'B-LOC', 'I-LOC']
</pre><p> </p><p>Again thanks for your time and feel free to reach out to me at any time if you have any further questions or comments.</p><p>This open-source contributing is becoming more and more interesting...</p>